File: Programmers/[67259] 경주로 건설/KingDonggyu/67259.js
function solution(board) {
  const size = board.length; 
  // 상, 우, 하, 좌
  const dx = [-1, 0, 1, 0];
  const dy = [0, 1, 0, -1];

  const visited = new Array(size).fill(null)
    .map((_) => new Array(size).fill(0));

  const queue = [[0, 0, -1, 0, 1]];

  while (queue.length > 0) {
    // (x, y) 좌표, 현재 방향, 누적 비용, 기회 횟수
    let [x, y, nowDi, cost, chance] = queue.shift();

    // 불필요한 연산을 줄이기 위해, 현재 방향을 우선순위로 한다
    const di = [nowDi < 0 ? 1 : nowDi];
    for (let i = 0; i < 4; i++) {
      if (i === di[0]) continue;
      di.push(i);
    }

    for (const nextDi of di) {
      const nextX = x + dx[nextDi];
      const nextY = y + dy[nextDi];

      if (nextX < 0 || nextX >= size || nextY < 0 || nextY >= size) {
        continue;
      }

      if (!board[nextX][nextY]) {
        const nextCost =
          nowDi < 0 || nowDi === nextDi ? cost + 100 : cost + 600;
          
        if (visited[nextX][nextY] && nextCost > visited[nextX][nextY]) {
          // 다음 이동에서 최소값을 기대할 수 있기 때문에, 기회를 준다
          if (chance === 1) {
            queue.push([nextX, nextY, nextDi, nextCost, chance - 1]);
          }
          continue;
        }
        visited[nextX][nextY] = nextCost;
        queue.push([nextX, nextY, nextDi, nextCost, chance]);
      }
    }
  }

  return visited[size - 1][size - 1];
}

File: Programmers/[67259] 경주로 건설/JFe/67259.py
from collections import deque
import sys

def solution(board):
    answer, N = 0, len(board)
    dx = [1, 0, -1, 0]
    dy = [0, 1, 0, -1]
    cost_list = [[[sys.maxsize for _ in range(4)]for _ in range(N)]for _ in range(N)]   # y좌표, x좌표, 방향에 해당하는 누적 최소 비용 저장
    x, y, d, cost = 0, 0, -1, 0
    # BFS 탐색
    q = deque()
    q.append((x, y, d, cost))
    while q:
        x, y, d, cost = q.popleft()
        for i in range(4):
            next_x, next_y = x + dx[i], y + dy[i]
            # boundary 안이고 벽이 아닌 경우
            if 0 <= next_x < N and 0 <= next_y < N and board[next_y][next_x] == 0:
                # 직선 도로 비용 추가
                next_cost = cost + 100
                # 방향이 달라지면 코너 비용 추가
                if d != i and d != -1:
                    next_cost += 500
                # 해당 cost_list 값보다 작으면 업데이트 하고 append
                if next_cost < cost_list[next_y][next_x][i]:
                    cost_list[next_y][next_x][i] = next_cost
                    q.append((next_x, next_y, i, next_cost))
    answer = min(cost_list[N-1][N-1])
    return answer